src/compiler/checker.ts

```diff
@@ -10269,7 +10269,7 @@ namespace ts {
                 !(prop.flags & SymbolFlags.Method && propType.flags & TypeFlags.Union)) {
                 return propType;
             }
-            return getFlowTypeOfReference(node, propType, /*assumeInitialized*/ true, /*includeOuterFunctions*/ false);
+            return getFlowTypeOfReference(node, propType, /*assumeInitialized*/ true, /*includeOuterFunctions*/ isReadonlySymbol(prop));

```

Comment:
I think you need a stronger check here. Say that you have a property access for `x.y.z`. Right now you're just checking whether `z` is a readonly symbol, but you really need to check for `x.y` and `x` as well.